integration/e2e_test.go

```diff
@@ -299,3 +302,89 @@ func deployMicrobot(t testing.TB) {
 		}
 	}
 }
+
+func upgradeCluster(t testing.TB) {
+	version := os.Getenv("CAPI_UPGRADE_VERSION")
+	if version == "" {
+		t.Fatalf("Environment variable CAPI_UPGRADE_VERSION is not set." +
+			"Please set it to the version you want to upgrade to.")
+	}
+
+	control_plane_name := os.Getenv("CAPI_UPGRADE_CP_NAME")
+	if control_plane_name == "" {
+		t.Fatalf("Environment variable CAPI_UPGRADE_CP_NAME is not set." +
+			"Please set it to the name of the control plane you want to upgrade.")
+	}
+
+	control_plane_type := os.Getenv("CAPI_UPGRADE_CP_TYPE")
+	if control_plane_type == "" {
+		t.Fatalf("Environment variable CAPI_UPGRADE_CP_TYPE is not set." +
+			"Please set it to the type of the control plane you want to upgrade.")
+	}
+
+	worker_deployment_name := os.Getenv("CAPI_UPGRADE_MD_NAME")
+	if worker_deployment_name == "" {
+		t.Fatalf("Environment variable CAPI_UPGRADE_MD_NAME is not set." +
+			"Please set it to the name of the machine deployment you want to upgrade.")
+	}
+
+	worker_deployment_type := os.Getenv("CAPI_UPGRADE_MD_TYPE")
+	if worker_deployment_type == "" {
+		t.Fatalf("Environment variable CAPI_UPGRADE_MD_TYPE is not set." +
+			"Please set it to the type of the machine deployment you want to upgrade.")
+	}
+
+	t.Logf("Upgrading cluster to %s", version)
+	// Patch contol plane machine upgrades.
+	patch := fmt.Sprintf(`{"spec":{"version":"%s"}}`, version)
+	command := []string{"kubectl", "patch", "--type=merge", control_plane_type, control_plane_name, "--patch", patch}
+	cmd := exec.Command(command[0], command[1:]...)
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		t.Error(string(output))
+		t.Fatalf("Failed to merge the patch to control plane. %s", err)
+	}
+
+	// Patch worker machine upgrades.
+	patch = fmt.Sprintf(`{"spec":{"template":{"spec":{"version":"%s"}}}}`, version)
+	command = []string{"kubectl", "patch", "--type=merge", worker_deployment_type, worker_deployment_name, "--patch", patch}
+	cmd = exec.Command(command[0], command[1:]...)
+	output, err = cmd.CombinedOutput()
+	if err != nil {
+		t.Error(string(output))
+		t.Fatalf("Failed to merge the patch to the machine deployments. %s", err)
+	}
+
+	time.Sleep(30 * time.Second)
+
+	// Now all the machines should be upgraded to the new version.
+	attempt := 0
+	max_attempts := 60
+	command = strings.Fields("kubectl get machine --no-headers")
+	for {
+		cmd := exec.Command(command[0], command[1:]...)
+		output, err := cmd.CombinedOutput()
+		output_str := string(output)
```

Comment:
nit; rename `output` to `outputBytes` and `output_str` to `output`

`output_str` is the variable you reuse a lot below, so make it have as less "context" in the name as possible.

Reply:
also below, prefer `lowerCamelCase` for variables.